src/operator/subgraph/build_subgraph.cc

```diff
@@ -329,7 +368,21 @@ void PreSelectSubgraphNodes(const nnvm::Graph& g, SubgraphSelectorV2Ptr subgraph
     }
     ++count;
   }
-  if (!success) {
+  if (success) {
+    // CachedOp requires external input during forward pass
+    // check subgraph input. If none, reject the first op (in top order) from the subgraph
+    // to make sure CachedOp gets external input.
+    // this feature can be switched off by setting ensure_CachedOp_input to false
+    const SubgraphPropertyPtr& subg_prop = g.GetAttr<SubgraphPropertyPtr>("subgraph_property");
+    if (subg_prop->HasAttr("ensure_CachedOp_input")
+        && subg_prop->GetAttr<bool>("ensure_CachedOp_input")) {
+      if (subgraph_nodes->size() > 0 && !HasInputEntries(g, simple_nodes, *subgraph_nodes)) {
+        // relabel the node to -1
+        (*subgraph_nodes)[0]->label = -1;
+        subgraph_nodes->erase(subgraph_nodes->begin());
```

Comment:
Can this check be done inside the `Filter` API of the subgraph property? And then reject the single node if needed there?

Reply:
here we wanted to check if cachedOp has external input, which requires both the entire graph and the subgraph. 
Seems it’s not trivial to move this functionality to `filter` as it is not easy to pass the info to there.

Reply:
If we keep this logic here, then it can also be used by other subgraph properties by setting `ensure_CachedOp_input` flag.

Reply:
sounds good, can we rename the flag to something more general like`requireSubgraphInputs`?